ledger/btcMessageSigning.ts

```diff
@@ -0,0 +1,159 @@
+import AppClient, { DefaultWalletPolicy } from 'ledger-bitcoin';
+import { NetworkType } from '../types';
+import { Bip32Derivation, TapBip32Derivation, Transport } from './types';
+import { getCoinType, getNativeSegwitAccountDataFromXpub, getTaprootAccountDataFromXpub } from './helper';
+import { AddressType, getAddressInfo } from 'bitcoin-address-validation';
+import { BTC_SEGWIT_PATH_PURPOSE, BTC_TAPROOT_PATH_PURPOSE } from '../constant';
+import { encode } from 'varuint-bitcoin';
+
+import { bip0322Hash } from '../connect';
+import { Psbt, Transaction } from 'bitcoinjs-lib';
+
+/**
+ * This function is used to sign an incoming BIP 322 message with the ledger
+ * @param transport - the transport object with connected ledger device
+ * @param networkType - the network type (Mainnet or Testnet)
+ * @param addressIndex - the index of the account address to sign with
+ * @param address - the address used to sign the message
+ * @param message - the incoming message in string format to sign
+ * @returns the signature in string (base64) format
+ * */
+export async function signSimpleBip322Message({
+  transport,
+  networkType,
+  addressIndex,
+  address,
+  message,
+}: {
+  transport: Transport;
+  networkType: NetworkType;
+  addressIndex: number;
+  address: string;
+  message: string;
+}) {
+  const app = new AppClient(transport);
+  const coinType = getCoinType(networkType);
+  const { type } = getAddressInfo(address);
+  const prevoutHash = Buffer.from('0000000000000000000000000000000000000000000000000000000000000000', 'hex');
+  const prevoutIndex = 0xffffffff;
+  const sequence = 0;
+  const scriptSig = Buffer.concat([Buffer.from('0020', 'hex'), Buffer.from(bip0322Hash(message), 'hex')]);
+  // Get account details from ledger to not rely on state
+  const masterFingerPrint = await app.getMasterFingerprint();
+  if (type === AddressType.p2tr) {
```

Comment:
This function is huge. Could you please extract 2 more functions for the taproot and segwit signing and call those from within the ifs?

Reply:
There's actually a lot of repetition between the 2 implementations. Could we maybe even extract a single function that we call with different inputs depending on the address type that does this in 1 shared bit of code?